### PR TITLE
[Hunt Tuning] Enforce `STATS` or `KEEP` functions in ES|QL hunting queries

### PR DESCRIPTION
**Pull request by terrancedejesus**
# Pull Request

*Issue link(s)*:
* https://github.com/elastic/detection-rules/pull/4146

## Summary - What I changed

* Updated all queries that did not have `keep` or `stats` in them to include at least `keep`
* Added `validate_esql_query` method to `Hunt` dataclass that is called only in a query starts with `from` (assuming all ES|QL queries for hunting will use `FROM` source commands only

## How To Test
Run: `pytest tests/test_hunt_data.py` from this branch to check for success. Then make adjustments to any hunt queries that are ES|QL and remove the Keep or Stats functions.

Screenshot 2024-10-15 at 7 18 19 PM

## Checklist

- [x] Added a label for the type of pr: `bug`, `enhancement`, `schema`, `Rule: New`, `Rule: Deprecation`, `Rule: Tuning`, `Hunt: New`, or `Hunt: Tuning` so guidelines can be generated
- [x] Added the `meta:rapid-merge` label if planning to merge within 24 hours
- [ ] Secret and sensitive material has been managed correctly
- [ ] Automated testing was updated or added to match the most common scenarios
- [ ] Documentation and comments were added for features that require explanation

## Contributor checklist

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/detection-rules/blob/main/CONTRIBUTING.md)?

### CODE REVIEW

**Comment by protectionsmachine** [comment]
## Hunt: Tuning - Guidelines

These guidelines serve as a reminder set of considerations when tuning an existing Hunt.

### Documentation and Context

- [ ] Detailed description of the suggested changes.
- [ ] Provide example JSON data or screenshots.
- [ ] Evidence of enhancing hunting results by either reducing false-positives or removing false-negatives.
- [ ] Evidence of specific environment factors influencing customized hunt tuning (Contextual Tuning).
- [ ] Evidence of refining hunts to better detect deviations from typical behavior (Behavioral Tuning).
- [ ] Field Usage: Ensure standardized fields for compatibility across different data environments and sources.

### Hunt Metadata Checks

- [ ] `author`: The name of the individual or organization authoring the rule.
- [ ] `name` and `description` are descriptive and typo-free.
- [ ] `language`: The query language(s) used in the rule, such as `KQL`, `EQL`, `ES|QL`, `OsQuery`, or `YARA`.
- [ ] `query` is inclusive, not overly exclusive. Review to ensure the original intent of the hunt is maintained.
- [ ] `integration` aligns with the `index`. Ensure updates if the integration is newly introduced.
- [ ] `notes` includes additional information (e.g., Triage and analysis investigation guides, timeline templates).
- [ ] `mitre` matches appropriate technique and sub-technique IDs that hunting query collect's data for.
- [ ] `references` are valid URL links that include information relevenat to the hunt or threat.

### Testing and Validation

- [ ] Evidence of testing and valid query usage.
- [ ] Markdown Generated: Run `python -m hunting generate-markdown` with specific parameters to ensure a markdown version of the hunting TOML files is created.
- [ ] Index Refreshed: Run `python -m hunting refresh-index` to refresh indexes.
- [ ] Run Unit Tests: Run `pytest tests/test_hunt_data.py` to run unit tests.

File: hunting/definitions.py [hunk 1]
```diff
@@ -22,15 +23,43 @@
 
 @dataclass
 class Hunt:
-    """Dataclass to represent a hunt."""
+    """Base class to represent a hunt."""
```
Comment (terrancedejesus) [inline hunting/definitions.py:26]:
```suggestion
    """Dataclass to represent a hunt."""
```

File: hunting/definitions.py [hunk 2]
```diff
@@ -25,12 +26,40 @@ class Hunt:
     """Dataclass to represent a hunt."""
     author: str
     description: str
-    integration: list[str]
+    integration: List[str]
```
Comment (terrancedejesus) [inline hunting/definitions.py:29]:
```suggestion
    integration: list[str]
```

File: hunting/definitions.py [hunk 3]
```diff
@@ -28,9 +29,37 @@ class Hunt:
     integration: list[str]
     uuid: str
     name: str
-    language: list[str]
+    language: List[str]
```
Comment (terrancedejesus) [inline hunting/definitions.py:32]:
```suggestion
    language: list[str]
```

**Review by Mikaayenson — APPROVED** [review]
May want to update the best practices to mention both keep and stats for optimization purposes. Otherwise lgtm.

**Comment by shashank-elastic** [comment]
this should be enforced in llm hunting as well right?, If so happy to add those in the same PR

**Comment by terrancedejesus** [comment]
> this should be enforced in llm hunting as well right?, If so happy to add those in the same PR

@shashank-elastic the LLM hunts are fine as the logic requires STATS or KEEP. Looking at those LLM hunts, they are using STATS.

File: hunting/definitions.py [hunk 4]
```diff
@@ -25,12 +26,40 @@ class Hunt:
...
+
+        if self.author == "Elastic":
+            # Regex patterns for checking "stats by" and "| keep"
+            stats_by_pattern = re.compile(r'\bstats\b.*?\bby\b', re.DOTALL)
+            keep_pattern = re.compile(r'\| keep', re.DOTALL)
+
+            # Check if either "stats by" or "| keep" exists in the query
+            if not stats_by_pattern.search(query) and not keep_pattern.search(query):
+                raise ValueError(
+                    f"Hunt: {self.name} contains an ES|QL query that must contain either 'stats by' or '| keep'."
```
Comment (terrancedejesus) [inline hunting/definitions.py:64]:
```suggestion
                    f"Hunt: {self.name} contains an ES|QL query that must contain either 'stats by' or 'keep' functions."
```

File: hunting/definitions.py [hunk 5]
```diff
@@ -61,5 +61,5 @@ def validate_esql_query(self, query: str) -> None:
             # Check if either "stats by" or "| keep" exists in the query
             if not stats_by_pattern.search(query) and not keep_pattern.search(query):
                 raise ValueError(
-                    f"Hunt: {self.name} contains an ES|QL query that must contain either 'stats by' or '| keep'."
+                    f"Hunt: {self.name} contains an ES|QL query that must contain either 'stats by' or 'keep' functions."
```
Comment (shashank-elastic) [inline hunting/definitions.py:64]: Perfect!

**Review by shashank-elastic — APPROVED** [review]

**Pull request merged**